crates/nix_rs/src/store.rs

```diff
@@ -147,38 +147,41 @@ impl NixStoreCmd {
         &self,
         out_paths: HashSet<StorePath>,
     ) -> Result<HashSet<StorePath>, NixStoreCmdError> {
-        let mut all_drvs = HashSet::new();
-        for out in out_paths.iter() {
-            let drv = self.nix_store_query_deriver(out.clone()).await?;
-            all_drvs.insert(drv);
-        }
-        let mut all_outs = HashSet::new();
-        for drv in all_drvs {
-            let deps = self.nix_store_query_requisites_with_outputs(drv).await?;
-            all_outs.extend(deps.into_iter());
-        }
+        let all_drvs = self
+            .nix_store_query_deriver(&out_paths.iter().cloned().collect::<Vec<_>>())
+            .await?;
+        let all_outs = self
+            .nix_store_query_requisites_with_outputs(&all_drvs.iter().cloned().collect::<Vec<_>>())
+            .await?;
         Ok(all_outs)
     }
 
-    /// Return the derivation used to build the given build output.
+    /// Return the derivations used to build the given build output.
     pub async fn nix_store_query_deriver(
         &self,
-        out_path: StorePath,
-    ) -> Result<PathBuf, NixStoreCmdError> {
+        out_paths: &[StorePath],
+    ) -> Result<HashSet<PathBuf>, NixStoreCmdError> {
         let mut cmd = self.command();
-        cmd.args([
-            "--query",
-            "--valid-derivers",
-            out_path.as_path().to_string_lossy().as_ref(),
-        ]);
+        let mut paths = vec![];
+        for path in out_paths {
+            paths.push(path.as_path());
+        }
+        cmd.args(["--query", "--valid-derivers"]).args(paths);
+
         crate::command::trace_cmd(&cmd);
         let out = cmd.output().await?;
         if out.status.success() {
-            let drv_path = String::from_utf8(out.stdout)?.trim().to_string();
-            if drv_path == "unknown-deriver" {
+            let out = String::from_utf8(out.stdout)?.trim().to_string();
+            let drv_paths: HashSet<PathBuf> = out
+                .lines()
+                .map(|l| l.trim().to_string())
+                .filter(|l| !l.is_empty())
```

Comment:
Why do we need these two lines? What happens if you get rid of them?

Reply:
You're right. These are actually not needed as we have applied trim() before to the whole String so chances of getting whitespace or empty lines won't be there. I can safely remove these two extra lines.

Reply:
I tested it, it behaves normally.